pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -100,15 +104,35 @@ func setPodInterfaceCache(iface *v1.Interface, podInterfaceName string, uid stri
 		return nil
 	}
 
+	cache := PodCacheInterface{Iface: iface}
+	indexIPv4 := indexNA
+	indexIPv6 := indexNA
+
 	for _, addr := range addrList {
 		if addr.IP.IsGlobalUnicast() {
-			err = writeToCachedFile(PodCacheInterface{Iface: iface, PodIP: addr.IP.String()},
-				util.VMIInterfacepath, uid, iface.Name)
-			if err != nil {
-				log.Log.Reason(err).Errorf("failed to write pod Interface to cache, %s", err.Error())
-				return err
+			addrString := addr.IP.String()
+			if netutils.IsIPv6(addr.IP) {
+				if indexIPv6 == indexNA {
+					indexIPv6 = len(cache.PodIPs)
+					cache.PodIPs = append(cache.PodIPs, addrString)
+				}
+			} else if indexIPv4 == indexNA {
+				indexIPv4 = len(cache.PodIPs)
+				cache.PodIPs = append(cache.PodIPs, addrString)
 			}
-			return nil
+		}
+	}
+
+	if Handler.IsIpv4Primary() && indexIPv6 == 0 {
+		cache.PodIPs = []string{cache.PodIPs[1], cache.PodIPs[0]}
+	}
```

Comment:
Why don't you just copy the list of IPs from the Pood to the cache? Is it due to the bug of "IPv6 can be never primary" that is *not* documented in the API?

Is there a precedent from another codebase where somebody would try to recover from clearly wrongly configured cluster?

Right now, I don't see a reason for any special handling. The only supported order is IPv4 first. If the cluster is properly deployed, there will be no problem.

Reply:
> Why don't you just copy the list of IPs from the Pood to the cache? Is it due to the bug of "IPv6 can be never primary" that is _not_ documented in the API?

I used the logic that was there already, which get the IPs according 
Handler.AddrList(link, netlink.FAMILY_ALL)
and extended it to support dual IPs

> 
> Is there a precedent from another codebase where somebody would try to recover from clearly wrongly configured cluster?

Please give an example of which wrong config you mean if possible.
(maybe the next inline will shed will answer a bit)

> 
> Right now, I don't see a reason for any special handling. The only supported order is IPv4 first. If the cluster is properly deployed, there will be no problem.

You asked in previous thread here that we will auto detect what is the primary IP, and quote a section from k8s, which this code obey:
1. it holds max 1 IP per IP family.
2. PodIP == IPs[0]

Reply:
> > Why don't you just copy the list of IPs from the Pood to the cache? Is it due to the bug of "IPv6 can be never primary" that is _not_ documented in the API?
> 
> I used the logic that was there already, which get the IPs according
> Handler.AddrList(link, netlink.FAMILY_ALL)
> and extended it to support dual IPs
> https://github.com/kubevirt/kubevirt/pull/3958/files#diff-d3ed8eb8cca020f37cab207c079f2627R97
> `addrList, err := Handler.AddrList(link, netlink.FAMILY_ALL)` (in case the link doesnt work)
> Its a continuation of what we have, which is straight forward, without need of API calls, and to get the client.

That does not answer my question.

I see that we are building the list of IPs based on MY_POD_IP and IPs found using netlink. Missed that before, disregard my following question.

> 
> > Is there a precedent from another codebase where somebody would try to recover from clearly wrongly configured cluster?
> 
> Please give an example of which wrong config you mean if possible.
> (maybe the next inline will add more info, as the solution cover all the cases, beside empty list, that was there before as well - no regression, in case of empty list, the VM wont start - as was before, because the file wont be created, and is mandatory)
> 
> > Right now, I don't see a reason for any special handling. The only supported order is IPv4 first. If the cluster is properly deployed, there will be no problem.
> 
> You asked in previous thread [2] in this PR that we will auto detect what is the primary IP, and quote [1] a section from k8s, which this code obey:

No. I asked for the opposite. I said we should expect Kubernetes to report it in the correct order, so we don't need to detect anything.

> 
>     1. it holds max 1 IP per IP family.
> 
>     2. PodIP == IPs[0] - so in the future if IPv6 will be primary we will already support it.
> 
> 
> [1] [#3958 (comment)](https://github.com/kubevirt/kubevirt/pull/3958#discussion_r467895857)
> [2] [#3958 (comment)](https://github.com/kubevirt/kubevirt/pull/3958#discussion_r467913680)

The use of helper variables (index) and the switch of PodIPs if IPv4 is primary but IPv6 was found first is hardly readable. Make the code clearer. Even something as simple as following would make it far more readable:

```
get IPv4 if any
get IPv6 if any
if the default is IPv4, return IPv4 with optional IPv6
the opposite
```

Reply:
imo it will be less nice, because there will be more conditions, as there are 5 cases that need to be covered, 
i will add a commit with it, that we can discuss

Reply:
done